### PR TITLE
Optimize bitpacked `take`

### PR DESCRIPTION
**Pull request by lwwmanning**
todos:

- [x] write a benchmark for bitpacked `take`
- [x] figure out the optimal threshold for taking single values vs decompressing the chunk
- [x] resolve conflicts with #189 and use reinterpret_cast
- [x] tests to verify that we correctly handle patching & validity
- [x] handle patching & validity

### CODE REVIEW

File: vortex-fastlanes/src/bitpacking/compute.rs [hunk 1]
```diff
@@ -52,31 +54,48 @@ impl ScalarAtFn for BitPackedArray {
...
+
+        // Group indices into 1024-element chunks and relativise them to the beginning of each chunk
         let relative_indices: Vec<(usize, Vec<u16>)> = match_each_integer_ptype!(prim_indices.ptype(), |$P| {
-            let grouped_indices = prim_indices
+            prim_indices
                 .typed_data::<$P>()
                 .iter()
-                .group_by(|idx| (**idx / 1024) as usize);
-            grouped_indices
+                .sorted()
```
Comment (robert3005) [inline vortex-fastlanes/src/bitpacking/compute.rs]: This won't work - you have to return results in the same order they have been asked for

Comment (robert3005) [reply vortex-fastlanes/src/bitpacking/compute.rs]: so as you deduplicate you need to create another list of indicies in the deduplicated list. i.e. if you have `[2, 2000, 2, 2000]` I'd produce something like `[(0, [2]), (1, [976])]` and `[(0, 0), (1, 0), (0, 0), (1, 0)]`.

File: vortex-fastlanes/src/bitpacking/compute.rs [hunk 2]
```diff
@@ -52,31 +54,48 @@ impl ScalarAtFn for BitPackedArray {
...
+    size_hint: usize,
+) -> VortexResult<Vec<T>> {
+    let mut output = Vec::with_capacity(size_hint);
+    for (chunk, offsets) in relative_indices {
+        let sliced = array.slice(chunk * 1024, (chunk + 1) * 1024).unwrap();
+        let sliced = sliced.as_bitpacked();
+        let packed = flatten_primitive(sliced.encoded()).unwrap();
+        let packed = packed.typed_data::<u8>();
+
+        // TODO(wmanning): if offsets.len() over threshold, unpack the whole thing
```
Comment (robert3005) [inline vortex-fastlanes/src/bitpacking/compute.rs]: you can avoid slicing first if you're below len and just take `chunk * 1024 + index`

File: vortex-fastlanes/src/bitpacking/compute.rs (hunk 1 above)
Comment (lwwmanning) [reply vortex-fastlanes/src/bitpacking/compute.rs]: I wonder if that's worth the complexity, or if we just accept that out of order takes will just go down the "singles" path?

File: vortex-fastlanes/src/bitpacking/compress.rs [hunk 3]
```diff
@@ -101,17 +101,13 @@ impl EncodingCompression for BitPackedEncoding {
...
     // TODO(ngates): we should implement this using a vortex cast to centralize this hack.
-    let bytes = match parray.ptype() {
-        I8 | U8 => bitpack_primitive(parray.buffer().typed_data::<u8>(), bit_width),
-        I16 | U16 => bitpack_primitive(parray.buffer().typed_data::<u16>(), bit_width),
-        I32 | U32 => bitpack_primitive(parray.buffer().typed_data::<u32>(), bit_width),
-        I64 | U64 => bitpack_primitive(parray.buffer().typed_data::<u64>(), bit_width),
-        _ => panic!("Unsupported ptype {}", parray.ptype()),
-    };
-    PrimitiveArray::from(bytes).into_array()
+    let bytes = match_integers_by_width!(parray.ptype(), |$P| {
```
Comment (robert3005) [inline vortex-fastlanes/src/bitpacking/compress.rs:107]: I was looking at this code earlier today... might be easier to reinterpret cast using the function I just added in #189 and then just match on unsigned? But we'd have to add ptype.to_signed/to_unsigned

File: vortex-fastlanes/src/bitpacking/compute.rs [hunk 4]
```diff
@@ -51,39 +54,101 @@ impl ScalarAtFn for BitPackedArray {
...
+        } else {
+            for index in offsets {
+                output.push(unsafe {
+                    unpack_single_primitive::<T>(packed_chunk, bit_width, index as usize)?
+                });
+            }
+        }
+    }
+
+    if let Some(patches) = array.patches() {
```
Comment (robert3005) [inline vortex-fastlanes/src/bitpacking/compute.rs]: After #206 lands you could taken on patches and then PrimitiveArray::patch?

**Comment by lwwmanning** [comment]
Screenshot 2024-04-05 at 12 57 10

As of https://github.com/fulcrum-so/vortex/pull/192/commits/b06f4023f0021d45ddbdacb45dd0553b5103b80a we have a dramatic improvement over develop (https://github.com/fulcrum-so/vortex/commit/0496b548b71979772a7b781281acc3277414657c) for unpatched arrays, and for 10K random indices on patched arrays. We see a bit of a slowdown on contiguous 10K take (of the same 100 indices, repeated 100 times) of patched arrays, going to investigate why.

**Comment by robert3005** [comment]
My only guess is that blindly patching with sliced sparse array is faster than first taking from that array

**Comment by lwwmanning** [comment]
Screenshot 2024-04-05 at 17 31 45

final benchmarks of commit https://github.com/fulcrum-so/vortex/pull/192/commits/f1bccc13435dd98ab93a5c650907362ece76461b against the implementation on develop from my M2 MacBook Air

**Comment by lwwmanning** [comment]
Implementing #207 should speed things up a bit more as well

File: vortex-fastlanes/src/lib.rs [hunk 5]
```diff
@@ -3,7 +3,8 @@
 
 pub use bitpacking::*;
 pub use delta::*;
-use linkme::distributed_slice;
+pub use downcast::*;
+pub use linkme::distributed_slice;
```
Comment (robert3005) [inline vortex-fastlanes/src/lib.rs]:
```suggestion
use linkme::distributed_slice;
```
This shouldn’t be pub

**Review by robert3005 — APPROVED** [review]

**Pull request merged**